apollo-gradle-plugin-incubating/src/main/kotlin/com/apollographql/apollo/gradle/kotlin/ApolloPlugin.kt

```diff
@@ -0,0 +1,247 @@
+package com.apollographql.apollo.gradle.kotlin
+
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.internal.tasks.factory.dependsOn
+import com.apollographql.apollo.compiler.GraphQLCompiler
+import org.gradle.api.Action
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.internal.HasConvention
+import org.gradle.api.internal.file.FileResolver
+import java.io.File
+import org.gradle.api.plugins.JavaPluginConvention
+import org.gradle.api.tasks.TaskProvider
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
+import javax.inject.Inject
+import kotlin.IllegalArgumentException
+
+
+class ApolloPlugin : Plugin<Project> {
+  companion object {
+    const val TASK_GROUP = "apollo"
+  }
+
+  lateinit var apolloExtension: ApolloExtension
+
+  override fun apply(project: Project) {
+    apolloExtension = project.extensions.create("apollo", ApolloExtension::class.java)
+
+    // the extension block has not been evaluated yet, register a callback once the project has been evaluated
+    project.afterEvaluate {
```

Comment:
This looks like a drawback compared to previous implementation. Would it be possible to use `project.plugins.withId` or similar dynamic callbacks to trigger the configuration more reactively based on other applied projects?

Reply:
I'm not sure about that. See https://github.com/apollographql/apollo-android/pull/1507#discussion_r314950718. 

We could use properties and postpone most of the extension resolution to the execution phase. The problem is the `generateKotlinModels` that requires to wire/setup tasks differently.

Despite all the lazy configuration and properties magic, it looks like everyone is still relying on `afterEvaluate` for these use cases...

Reply:
I'm not sure about that. See #1507 (comment).

We could use properties and postpone most of the extension resolution to the execution phase. The problem is the generateKotlinModels that requires to wire/setup tasks differently.

Despite all the lazy configuration and properties magic, it looks like everyone is still relying on afterEvaluate for these use cases...